pre-sync/oci-image-verification/main.go

```diff
@@ -0,0 +1,145 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os/exec"
+	"sync"
+
+	admissionv1 "k8s.io/api/admission/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	SourceURL    = "configsync.gke.io/source-url"
+	SourceCommit = "configsync.gke.io/source-commit"
+)
+
+var authorized bool
+var authMutex sync.Mutex // Add a mutex for thread safety
+
+// Function to extract annotations from JSON
+func getAnnotations(raw []byte) (map[string]string, error) {
+	var metadata map[string]interface{}
+	if err := json.Unmarshal(raw, &metadata); err != nil {
+		return nil, err
+	}
+	if annotations, ok := metadata["metadata"].(map[string]interface{})["annotations"].(map[string]interface{}); ok {
+		annotationsMap := make(map[string]string)
+		for k, v := range annotations {
+			annotationsMap[k] = fmt.Sprintf("%v", v)
+		}
+		return annotationsMap, nil
+	}
+	return nil, fmt.Errorf("no annotations found")
+}
+
+func validateImage(image string) error {
+	cmd := exec.Command("cosign", "verify", image, "--key", "/cosign-key/cosign.pub")
+	log.Printf("command %s", cmd.String())
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("cosign verification failed: %s, output: %s", err, string(output))
+	}
+	return nil
+}
+
+func auth() error {
+	authMutex.Lock()         // Acquire the lock before accessing shared resources
+	defer authMutex.Unlock() // Release the lock when the function exits
+
+	if authorized {
+		log.Printf("skip authorizing docker")
+		return nil
+	}
+	gcloudCmd := exec.Command("gcloud", "auth", "print-access-token")
+	accessTokenBytes, err := gcloudCmd.Output()
+	if err != nil {
+		return fmt.Errorf("Error fetching access token: %v\n", err)
+	}
+	// Convert the output to a string and trim any trailing newline
+	accessToken := string(accessTokenBytes)
+	accessToken = accessToken[:len(accessToken)-1] // Remove the trailing newline
+
+	cmd := exec.Command("cosign", "login", "us-central1-docker.pkg.dev", "-u", "oauth2accesstoken", "-p", accessToken)
```

Comment:
This adds another assumption that the image is stored in GAR with the `us-central1-docker` region. We would need to make it more generic to support other OCI registries.